### ISSUE TITLE
Reference: Architecture — system shape, abstractions, locales

### ISSUE THREAD

**Issue by GirlBossRush** [isp-nexus/mailwoman#4]
# Architecture

## System shape

```
┌─────────────────────────────────────────────────────────────────┐
│                         Mailwoman SDK                            │
│                                                                  │
│   ┌──────────────────┐   ┌────────────────────────────────┐     │
│   │  Tokenization    │──▶│  Classifiers (rule + neural)   │     │
│   │  (existing)      │   │                                │     │
│   └──────────────────┘   │  ┌──────────────────────────┐  │     │
│                          │  │ rule: house_number       │  │     │
│                          │  │ rule: postcode           │  │     │
│                          │  │ rule: street_prefix      │  │     │
│                          │  │ rule: whos_on_first      │  │     │
│                          │  │ ...                      │  │     │
│                          │  │                          │  │     │
│                          │  │ neural: NeuralSequence   │──┼──┐  │
│                          │  └──────────────────────────┘  │  │  │
│                          └────────────────────────────────┘  │  │
│                                       │                       │  │
│                                       ▼                       │  │
│                          ┌────────────────────────────────┐  │  │
│                          │  ClassifierPolicy (per-tag)    │  │  │
│                          └────────────────────────────────┘  │  │
│                                       │                       │  │
│                                       ▼                       │  │
│                          ┌────────────────────────────────┐  │  │
│                          │  Solver (existing)             │  │  │
│                          │  ExclusiveCartesianSolver +    │  │  │
│                          │  filters + augmenters          │  │  │
│                          └────────────────────────────────┘  │  │
│                                       │                       │  │
│                                       ▼                       │  │
│                          ┌────────────────────────────────┐  │  │
│                          │  Ranked Solutions              │  │  │
│                          └────────────────────────────────┘  │  │
│                                                               │  │
└───────────────────────────────────────────────────────────────┼──┘
                                                                │
                          ┌─────────────────────────────────────┴──┐
                          │  @mailwoman/neural (new package)       │
                          │                                        │
                          │  ┌──────────────────────────────────┐  │
                          │  │  NeuralSequenceClassifier        │  │
                          │  │  - lazy-loads ONNX model         │  │
                          │  │  - SentencePiece tokenizer       │  │
                          │  │  - emits ClassificationProposal  │  │
                          │  └──────────────────────────────────┘  │
                          │              │                          │
                          │              ▼                          │
                          │  ┌──────────────────────────────────┐  │
                          │  │  Weights packages (per locale)   │  │
                          │  │  @mailwoman/neural-weights-en-us │  │
                          │  │  @mailwoman/neural-weights-fr-fr │  │
                          │  │  @mailwoman/neural-weights-ja-jp │  │
                          │  └──────────────────────────────────┘  │
                          └────────────────────────────────────────┘
```

## Monorepo layout

```
packages/
  core/                      # existing — tokenization, span/section/phrase, classifier base
  classifiers/               # existing rule classifiers — pulled out of core for clarity
  neural/                    # NEW — ONNX runtime + SentencePiece + sequence classifier
  neural-weights-en-us/      # NEW — ONNX model weights for US English
  neural-weights-fr-fr/      # NEW — ONNX model weights for FR French
  corpus/                    # NEW — TS adapters, alignment, synthesis, Parquet output
  corpus-python/             # NEW — Python training pipeline, NOT published to npm
  studio/                    # FUTURE — Phase 5 web UI for human correction
sdk/                         # existing — public consumer API
server/                      # existing — HTTP service
cli.ts                       # existing
```

The split between `neural/` (runtime) and `neural-weights-*/` (data) is deliberate. Users running US-only deployments should not download French weights. Weights packages are versioned independently from the runtime.

### `ComponentTag`

The canonical union of address component types. Defined in #5. Single source of truth. Adding a tag requires a written rationale.

### `ClassificationProposal`

The shape every classifier produces. Replaces the current ad-hoc shape used by rule classifiers. See #6.

```ts
interface ClassificationProposal {
	span: Span
	component: ComponentTag
	confidence: number // 0..1
	source: "rule" | "neural" | "merged"
	source_id: string // 'house_number' for rule, 'neural-v0.3-en-us' for neural
	penalty: number
	metadata?: Record<string, unknown>
}
```

Existing rule classifiers wrap their output in this shape. Neural classifier emits the same shape. The solver does not distinguish.

### `ClassifierPolicy`

Per-component configuration that decides which classifier(s) get authority for that component. This is the Ship of Theseus dial.

```ts
interface ClassifierPolicy {
	component: ComponentTag
	mode: "rule_only" | "neural_only" | "both" | "neural_preferred" | "rule_preferred"
	confidence_threshold?: number
	locale?: string // policy can vary per locale
}
```

Default policy is `rule_only` for everything until neural is shipped. Migration happens one component at a time, gated on golden-set metrics.

### `LocaleProfile`

Encapsulates locale-specific behavior: which weights package to load, which rule classifiers apply, which synthesis rules govern training data generation.

```ts
interface LocaleProfile {
	locale: string // 'en-US', 'fr-FR', 'ja-JP'
	weightsPackage: string // npm package name
	ruleClassifiers: string[] // identifiers of rule classifiers active for this locale
	componentsSupported: ComponentTag[] // not every locale supports every tag (e.g. JP has no `street`)
	policy: ClassifierPolicy[]
}
```

## Locale strategy

Locales are first-class. The system does not assume English. Every classifier (rule or neural) declares which locales it serves.

- A US-only deployment loads `en-US` profile, English rule classifiers, US weights.
- A multi-locale deployment loads all configured profiles. Locale detection per input is a separate concern handled by a `LocaleDetector` (Phase 2 work; v0 can require explicit locale).
- New locale = new `LocaleProfile` + new weights package + (optionally) new rule classifiers. Core code does not change.

## Inference path

```
input string
     │
     ▼
Tokenization (existing Mailwoman code)
     │
     ▼
For each Section:
     │
     ├──▶ Rule classifiers produce ClassificationProposals
     │
     ├──▶ NeuralSequenceClassifier (if locale supports it):
     │       1. Tokenize section with SentencePiece (WASM)
     │       2. Run ONNX inference via onnxruntime-node
     │       3. Decode BIO labels back to character spans
     │       4. Emit ClassificationProposals
     │
     ▼
ClassifierPolicy filter:
     For each component, keep only proposals from authorized classifiers
     │
     ▼
Existing ExclusiveCartesianSolver
     │
     ▼
Ranked solutions
```

## Training path (Python, internal)

```
Source data (OSM, WOF, BAN, OpenAddresses, government registries)
     │
     ▼ TypeScript adapters (corpus/adapters/*.ts)
Canonical rows: { raw, components, country, source, source_id }
     │
     ▼ Alignment (corpus/align.ts)
BIO-labeled rows: { raw, tokens, labels, country, source, source_id }
     │
     ▼ Synthesis (corpus/synthesize.ts)
Augmented rows: + case variants, abbreviations, typos, ordering perturbations
     │
     ▼ Parquet shards (versioned)
corpus-vX.Y.Z/{train,val,test}-*.parquet
     │
     ▼ Python training (corpus-python/)
HuggingFace Transformers + PyTorch
     │
     ▼ ONNX export + int8 quantization
neural-weights-{locale}/model.onnx + tokenizer.model
     │
     ▼ Eval on golden set
If golden_F1 >= threshold: publish
```

## Why this shape

- **Classifier-as-plugin** preserves Mailwoman's existing extensibility. New rule classifiers, new neural variants, future learned rankers all conform to the same interface.
- **Per-component policy** allows incremental rollout without big-bang risk. A regression in neural `street` parsing doesn't affect `country` parsing.
- **Locale profiles** prevent the architecture from quietly Anglocentric-by-default. If the abstractions can't express a non-Anglo locale cleanly, we fix the abstractions, not the locale.
- **Separate weights packages** keep npm install times sane and let users opt in to languages they need.
- **Training in Python, inference in TS** uses each ecosystem for what it's best at and avoids the dual-language maintenance tax in production code.

## Anti-patterns to avoid

- ❌ A `mode: 'auto'` setting in `ClassifierPolicy` that "intelligently" picks rule vs neural. Hidden state, hard to debug. Always explicit.
- ❌ Hardcoding `'en-US'` as default in core. Default should be "no locale, classifiers must opt in."
- ❌ Loading all weights at module init. Lazy-load on first classification request.
- ❌ Mixing inference and resolution. The parser does not know about coordinates, place IDs, or WOF resolution. That's Phase 4.
- ❌ Adding tags to `ComponentTag` without updating the schema doc and the alignment logic in the same commit. Schema drift kills the corpus.

---
Part of epic #15.

**Comment by GirlBossRush** [comment]
**Operator note added to local ARCHITECTURE.md, 2026-05-17.** Mirroring here so it surfaces alongside the plan.

---

## Model architecture (what the neural classifier IS)

**Family: encoder-only transformer + per-token classification head.** Standard NER (named entity recognition) shape. Each token of the input address string gets exactly one BIO label from a fixed set (`O` + `B-<tag>` + `I-<tag>` for each tag in the `ComponentTag` union — currently 47 labels = 1 + 2 × 23 tags).

### Why this family, not the others

| Family | Used for | Why not us |
| --- | --- | --- |
| **Encoder-only + classification head** (BERT / RoBERTa / DeBERTa + NER heads) | Per-token tagging; same number of output labels as input tokens | **This is us.** Address parsing is a tagging problem, not a generation problem. |
| Encoder-decoder (T5, BART) | Seq2seq — translation, summarization | Output isn't a different sequence; it's structured tags per input token. Overkill + lossy for this shape. |
| Decoder-only (GPT, Llama) | Autoregressive generation | We annotate, we don't generate. Could approximate via constrained decoding at 1000× the cost. |
| Conditional random field (CRF) | Pre-transformer per-token tagging with Markov features | This is libpostal. We're replacing the CRF with a transformer encoder in front of an effectively-similar tag space. |

### Concrete config (Phase 2 / this issue's downstream #10)

- **Backbone**: HuggingFace `BertForTokenClassification` with custom small config — **6 layers, 256 hidden dim, 4 attention heads, 1024 FF intermediate, max position 128.** Trained **from scratch** (not fine-tuned from a pretrained English-internet BERT — the address vocabulary is too narrow and too unlike natural language for that pretraining to help).
- **Tokenizer**: SentencePiece (unigram), `vocab_size=16000`, `character_coverage=0.9995`, `byte_fallback=true`. Trained on the corpus, locked into corpus version.
- **Output head**: linear projection → 47 logits per token.
- **Inference**: softmax → argmax → BIO span decoding → `ClassificationProposal` per span with mean per-token softmax probability as confidence.
- **Size target**: < 30 MB int8 ONNX, < 40 MB including tokenizer. Few-million parameter count.
- **Runtime**: `onnxruntime-node`, CPU EP default, no Python at inference.

### The cousin: spaCy's `ja_core_news_trf`

The closest commercial reference point is spaCy's [Japanese transformer pipeline](https://spacy.io/models/ja#ja_core_news_trf):

| | spaCy `ja_core_news_trf` | mailwoman planned |
| --- | --- | --- |
| Backbone | `cl-tohoku/bert-base-japanese-char-v2` (BERT-base — 12 layers, 768d) | small BERT (6 layers, 256d) — ~10× smaller |
| Vocab | 6,144 char-level | 16,000 SentencePiece + byte fallback |
| Pretrained | Yes (Tohoku-Univ. Inui Lab BERT) | No (from scratch on address corpus) |
| Task heads | **3** on shared encoder: morphologizer (POS), parser (dependency), NER (22 entity types) | **1** on dedicated encoder: NER for 23 address component types |
| Model size | 320 MB | < 40 MB int8 |
| NER F1 (their domain) | 83.31 ENTS_F on general Japanese NER | TBD; the rule baseline is > 95% on `country` / `region` for the golden set, so the floor is high |
| Inference runtime | Python (spaCy) | Node (`onnxruntime-node`) |
| License | CC BY-SA 3.0 | AGPL-3.0 |

**spaCy validates the family choice** — when the industry-standard NER toolkit's flagship Japanese model is BERT-encoder + token-classification head, we're in the right shape. The trade-offs we make vs them:

- **Smaller** (10× fewer parameters): we're a narrower task. General-purpose NER must handle 22 entity types across all news text; address parsing handles 23 components in addresses only. Narrow task tolerates narrower model.
- **From scratch, not fine-tuned**: cl-tohoku was pretrained on Japanese Wikipedia + news. Our domain is far enough from natural-language text that pretraining transfer is weak; from-scratch on a domain-pure corpus gets us smaller, faster, more accurate.
- **Single head, not multi-head**: they share one encoder across morphologizer + parser + NER. We have one head only — encoder fully specializes for component disambiguation.

### Multi-head extension (Phase 4+)

The spaCy pipeline pattern — one shared encoder + multiple task heads — is a powerful future direction. Candidates if/when we want them:

- **Venue vs address-part head**: per-token binary classifier for compositional disambiguation (`Buffalo Health Clinic Buffalo NY` — is "Buffalo" venue-token or locality-token?). Useful for [the bitter-lesson kryptonite cases](https://github.com/sister-software/mailwoman/issues/3#issuecomment-4471520041).
- **Locale detection head**: predicts `en-US` vs `fr-FR` vs `ja-JP` for the whole input — lets `LocaleDetector` ride the same encoder.
- **Completeness head**: predicts whether the input is a complete address, a fragment, or noise — feeds graceful-failure signaling.
- **Confidence head**: a separate calibration learner, replacing the per-token softmax with a calibrated probability per the operator's "confidence-as-trap" framing.

None of these are Phase 0–3 scope. Filed for Phase 4 territory or beyond. spaCy's existence is the evidence the pattern works at scale.

### Casual descriptions (for audiences who don't want the architecture deep-dive)

- **Technical**: "Small encoder-only transformer (BERT-class, from scratch, ~few-million params, <40 MB int8 ONNX), per-token classification head emitting BIO-tagged address-component spans. Runs in Node via onnxruntime-node, no Python at inference."
- **Practitioner**: "Tiny BERT-class NER model specialized for address parsing — a learned replacement for libpostal's CRF, sized to fit in a Node process. Architecturally the same family as spaCy's ja_core_news_trf, ~10× smaller because the task is narrower."
- **User-facing**: "A small ONNX model that reads an address string and tells you, with per-component confidence, what each piece of it means."
- **For the bitter-lesson audience**: "The contextual-parser half of a contextual-parser-plus-constraint-solver geocoder front-end. The transformer proposes; the policy registry + solver dispose."

Notably NOT: a translator (no language pair, no autoregressive output). Not a generative model.

References:
- spaCy ja_core_news_trf: https://spacy.io/models/ja#ja_core_news_trf
- Backbone (Tohoku BERT-Japanese-char-v2): https://huggingface.co/cl-tohoku/bert-base-japanese-char-v2
- Release metadata (transformer config + label scheme + F1): https://github.com/explosion/spacy-models/releases/tag/ja_core_news_trf-3.7.0

**Comment by GirlBossRush** [comment]
**Additional operator note added to local ARCHITECTURE.md, 2026-05-17.** Mirroring here.

---

## Frameworks + tooling stance — what we use, what we don't, what we steal

**Use**: HuggingFace Transformers + PyTorch for training (Python-only, single-shot, output is the ONNX file). `onnxruntime-node` for inference (TypeScript / Node, the user-facing path). SentencePiece for tokenization on both sides. `@dsnp/parquetjs` (patched) for corpus output. That's the production stack — anything not on this list is either dead weight or duplicative.

**Don't use**: spaCy as a dependency, either runtime or training. Reasons:

- **Runtime**: spaCy is Python-only. The "TypeScript-first" stance is a hard constraint from the epic (#15); adding Python at inference time defeats the deployment-regression-for-Node-users argument that's the project's reason to exist. spaCy's ONNX export path (`spacy-onnx-export`) exists but is rarely production-grade.
- **Training**: spaCy's config-driven training is nice (`spacy.cfg` + `Thinc`) but pays off mostly when pipeline composition matters or the task is novel. Our task — single-head NER on `BertForTokenClassification` — is the most-trodden path in HuggingFace Transformers. spaCy adds an abstraction layer for marginal convenience. The operator's "less Python the better" stance argues against it even for training where Python is unavoidable.

**Steal the patterns, not the tool**:

| What | From | Where to apply |
| --- | --- | --- |
| Release notes template (label scheme table + transformer config block + per-component F1 + compatibility matrix) | spaCy's GitHub release format for models like [ja_core_news_trf-3.7.0](https://github.com/explosion/spacy-models/releases/tag/ja_core_news_trf-3.7.0) | Phase 3 (#11) `@mailwoman/neural-weights-*` package READMEs + model card JSON |
| Eval reporting conventions (per-component P/R/F1 + confusion matrix layout, `displaCy`-style visualization patterns) | spaCy benchmark reports | Phase 2 (#10) eval reports |
| "Training described entirely by a config file, code is just the runner" hygiene | spacy.cfg pattern | `corpus-python/train.py` config shape — adopt the pattern without taking spaCy as a dep |
| Vintage / language-model-version conventions | spaCy's `ja_core_news_trf-3.7.0` naming (`<lang>_<domain>_<size>_<framework>-<version>`) | Weights package versioning + ONNX file naming |

**Optional Phase 3 baseline comparison**: run a fine-tuned `xx_ent_wiki_sm` or `en_core_web_trf` against the same golden set, report comparison numbers in the model card. General NER tagging `LOC` / `ORG` where we expect `locality` / `venue` is informative — shows the address-specific gain over general NER. Operator-time, ~1 hour for an honest baseline run. Not in the install dependency tree; one-off comparison only.

**Comment by GirlBossRush** [comment]
Reference docs exist in docs/plan/reference/.
